File: packages/ref-sdk-py/alloist_ref_sdk/evidence.py
"""Evidence bundle verification per ACT-lite spec."""

from __future__ import annotations

import base64
import hashlib
import json
from typing import Any

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


def _canonical_json(obj: dict[str, Any]) -> bytes:
    return json.dumps(obj, sort_keys=True, separators=(",", ":")).encode("utf-8")


def verify_evidence_bundle(bundle: dict[str, Any]) -> bool:
    """
    Verify signed evidence bundle.
    Returns True if signature valid and input_hash (when present) matches.
    """
    data = dict(bundle)
    runtime_signature = data.pop("runtime_signature", None) or data.pop("runtimeSignature", None)
    public_key_b64 = data.pop("public_key", None) or data.pop("publicKey", None)

    if not runtime_signature or not public_key_b64:
        return False

    try:
        raw_pub = base64.b64decode(public_key_b64.encode("ascii"))
        public_key = Ed25519PublicKey.from_public_bytes(raw_pub)
        sig = base64.b64decode(runtime_signature.encode("ascii"))
    except Exception:
        return False

    payload_bytes = _canonical_json(data)
    try:
        public_key.verify(sig, payload_bytes)
    except Exception:
        return False

    input_hash = data.get("input_hash")
    if input_hash:
        excerpt = {
            "action_name": data.get("action_name", ""),
            "token_snapshot": data.get("token_snapshot", {}),
            "metadata": data.get("runtime_metadata", {}),
        }
        computed = hashlib.sha256(_canonical_json(excerpt)).hexdigest()
        if computed != input_hash:
            return False

    return True

File: packages/ref-sdk-node/src/evidence.js
'use strict';

const crypto = require('crypto');

function _canonicalJson(obj) {
  if (obj === null || typeof obj !== 'object') return JSON.stringify(obj);
  if (Array.isArray(obj)) return '[' + obj.map(_canonicalJson).join(',') + ']';
  const keys = Object.keys(obj).sort();
  const parts = keys.map((k) => JSON.stringify(k) + ':' + _canonicalJson(obj[k]));
  return '{' + parts.join(',') + '}';
}

function _toBase64Url(buf) {
  return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

/**
 * Verify signed evidence bundle per ACT-lite spec.
 * @param {object} bundle - Evidence bundle (must include runtime_signature, public_key)
 * @returns {boolean}
 */
function verifyEvidenceBundle(bundle) {
  const data = { ...bundle };
  const runtimeSignature = data.runtime_signature;
  const publicKeyB64 = data.public_key || data.publicKey;

  if (!runtimeSignature) return false;
  if (!publicKeyB64) return false;

  delete data.runtime_signature;
  delete data.runtimeSignature;
  delete data.public_key;
  delete data.publicKey;

  const payloadStr = _canonicalJson(data);
  const payloadBytes = Buffer.from(payloadStr, 'utf8');

  let publicKey;
  try {
    const raw = Buffer.from(publicKeyB64, 'base64');
    const x = _toBase64Url(raw);
    publicKey = crypto.createPublicKey({
      key: { kty: 'OKP', crv: 'Ed25519', x },
      format: 'jwk',
    });
  } catch {
    return false;
  }

  const sig = Buffer.from(runtimeSignature, 'base64');
  try {
    if (!crypto.verify(null, payloadBytes, publicKey, sig)) return false;
  } catch {
    return false;
  }

  const inputHash = data.input_hash;
  if (inputHash) {
    const excerpt = {
      action_name: data.action_name || '',
      token_snapshot: data.token_snapshot || {},
      metadata: data.runtime_metadata || {},
    };
    const computed = crypto.createHash('sha256').update(_canonicalJson(excerpt)).digest('hex');
    if (computed !== inputHash) return false;
  }

  return true;
}

module.exports = { verifyEvidenceBundle };
